jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/EditLambdaDialog.kt

```diff
@@ -0,0 +1,363 @@
+    handlerName: String = "",
+    envVariables: Map<String, String> = emptyMap(),
+    timeout: Int = DEFAULT_TIMEOUT,
+    role: IamRole? = null
+) : DialogWrapper(project) {
+
+    constructor(project: Project, lambdaFunction: LambdaFunction) :
+            this(
+                project = project,
+                isUpdate = true,
+                name = lambdaFunction.name,
+                description = lambdaFunction.description ?: "",
+                runtime = lambdaFunction.runtime,
+                handlerName = lambdaFunction.handler,
+                envVariables = lambdaFunction.envVariables ?: emptyMap(),
+                role = lambdaFunction.role
+            )
+
+    private val view = EditLambdaPanel(project)
+    private val validator = UploadToLambdaValidator()
+    private val s3Client: S3Client = AwsClientManager.getInstance(project).getClient()
+    private val iamClient: IamClient = AwsClientManager.getInstance(project).getClient()
+
+    private val updateAction = UpdateLambdaSettingsAction()
+    private val deployAction = DeployLambdaAction()
+
+    private var validateDeploySettings = false
+
+    init {
+        super.init()
+        title = if (isUpdate) message("lambda.upload.edit.title", name) else message("lambda.upload.create.title")
+
+        view.name.text = name
+
+        if (isUpdate) {
+            view.name.isEnabled = false
+        }
+
+        view.handler.text = handlerName
+        view.timeout.text = timeout.toString()
+        view.description.text = description
+        view.envVars.envVars = envVariables
+
+        view.sourceBucket.populateValues {
+            val activeRegionId = ProjectAccountSettingsManager.getInstance(project).activeRegion.id
+            s3Client.listBuckets().buckets()
+                .asSequence()
+                .mapNotNull { it?.name() }
+                .filter { s3Client.regionForBucket(it) == activeRegionId }
+                .sortedWith(String.CASE_INSENSITIVE_ORDER)
+                .toList()
+        }
+
+        view.createBucket.addActionListener {
+            val bucket = Messages.showInputDialog(
+                message("lambda.upload.create_s3_dialog.input"),
+                message("lambda.upload.create_s3_dialog.title"),
+                AwsIcons.Logos.S3_LARGE
+            )
+            bucket?.run {
+                view.sourceBucket.addAndSelectValue {
+                    s3Client.createBucket { request -> request.bucket(bucket) }
+                    bucket
+                }
+            }
+        }
+
+        view.runtime.populateValues(selected = runtime) { Runtime.knownValues() }
+
+        view.iamRole.populateValues(selected = role) {
+            iamClient.listRolesFilter { it.assumeRolePolicyDocument().contains(LAMBDA_PRINCIPAL) }
+                .map { IamRole(name = it.roleName(), arn = it.arn()) }
+                .sortedWith(Comparator.comparing<IamRole, String>(Function { it.name }, String.CASE_INSENSITIVE_ORDER))
```

Comment:
Disappointing that you need to do `toList` obviously by sorting it; there's no longer a `sequence` under the hood anymore!

Reply:
```kotlin
        override fun iterator(): Iterator<T> {
            val sortedList = this@sortedWith.toMutableList()
            sortedList.sortWith(comparator)
            return sortedList.iterator()
        }
```
Ya...super annoying